helm/config.go

```diff
@@ -0,0 +1,377 @@
+// the nodes will be encoded in the same order in which they were added to the object.
+type namedNode struct {
+	name string
+	node Node
+}
+
+// Object represents an ordered lst of named nodes.
+type Object struct {
+	sharedFields
+	nodes []namedNode
+}
+
+// NewObject creates an empty object node and initializes shared fields.
+func NewObject(modifiers ...NodeModifier) *Object {
+	object := &Object{}
+	object.Set(modifiers...)
+	return object
+}
+
+// Add a singled named node at the end of the list.
+func (object *Object) Add(name string, node Node) {
+	object.nodes = append(object.nodes, namedNode{name: name, node: node})
+}
+
+// Sort all nodes of the object by name.
+func (object *Object) Sort() {
+	sort.Slice(object.nodes[:], func(i, j int) bool {
+		return object.nodes[i].name < object.nodes[j].name
+	})
+}
+
+func (object Object) write(enc *Encoder, prefix string) {
+	for _, namedNode := range object.nodes {
+		enc.writeNode(namedNode.node, &prefix, namedNode.name+":")
+	}
+}
+
+// Encoder writes the config data to an output stream.
+type Encoder struct {
+	writer io.Writer
+
+	indent int
+	wrap   int
+
+	emptyLines     bool
+	pendingNewline bool
+}
+
+// EmptyLines turns generation of additional empty lines on or off. In general
+// each node that has a comment or a condition will be separated by additional
+// empty lines from the rest of the document. The leading empty line will be
+// omitted for the first element of a list or object, and the trailing empty
+// line will be omitted for the last element. The default value is false.
+func EmptyLines(emptyLines bool) func(*Encoder) {
+	return func(enc *Encoder) {
+		enc.emptyLines = emptyLines
+	}
+}
+
+// Indent sets the indentation amount per nesting level for the YAML encoding.
+// The default value is 2.
+func Indent(indent int) func(*Encoder) {
+	return func(enc *Encoder) {
+		if indent < 2 {
+			indent = 2
+		}
+		enc.indent = indent
+	}
+}
+
+// Wrap sets the maximum line length for comments. This number includes the
+// columns needed for indentation, so comments on more deeply nested nodes have
+// more tightly wrapped comments than outer level nodes. Wrapping applies only
+// to comments and not to conditions or scalar values. The default value is 80.
+func Wrap(wrap int) func(*Encoder) {
+	return func(enc *Encoder) {
+		enc.wrap = wrap
+	}
+}
+
+// Set applies Encoder modifier functions to set option values.
+func (enc *Encoder) Set(modifiers ...func(*Encoder)) {
+	for _, modifier := range modifiers {
+		modifier(enc)
+	}
+}
+
+// NewEncoder returns an Encoder object holding the output stream and encoding options.
+func NewEncoder(writer io.Writer, modifiers ...func(*Encoder)) *Encoder {
+	enc := &Encoder{
+		// Wrap io.Writer in a bufio.Writer so that we can check for
+		// errors at the very end by calling bufio.Writer.Flush()
+		writer:     bufio.NewWriter(writer),
+		emptyLines: false,
+		indent:     2,
```

Comment:
I kind of like the modifier functions because they can be passed through custom constructor wrappers (e.g. the `equal` function in the tests, or the `NewKubeConfig` utility function in the `kube` package.

They also allow for additional error checking, even though I don't use this right now (except to enforce that `indent >= 2`).

Reply:
Alright, mostly for the parameter passthrough thing.